mig/shared/base.py

```diff
@@ -505,23 +507,28 @@ def force_utf8(val, highlight=''):
         val = "%s" % val
     if not is_unicode(val):
         return val
-    return "%s%s%s" % (highlight, val.encode("utf8"), highlight)
+    return codecs.encode("%s%s%s" % (highlight, val, highlight), 'utf8')
+
+
+def _walk_and_covert_recursive(input_obj, highlight='', _as_bytes=False, _force_primitive=None, _force_recursive=None):
+    thetype = type(input_obj)
+    if issubclass(thetype, dict):
+        return {_force_recursive(i, highlight): _force_recursive(j, highlight) for (i, j) in
+                input_obj.items()}
+    elif issubclass(thetype, (list, tuple)):
+        return thetype((_force_recursive(i, highlight) for i in input_obj))
+    elif not is_unicode(input_obj):
```

Comment:
This `elif not is_unicode(input_obj):` does not look right to me when used from the `force_utf8_rec` case as it basically inverts the original logic there. I therefore suppose you'll end up in fact _not_ converting any unicode string entries to utf8 bytes, as we would really want with `_force_primitive` in that case. It _is_ correct when used from the `force_unicode_rec` case, though. Perhaps the seemingly unused `_as_bytes` argument was meant for that discrimination?

In fact it looks like it would work to modify the `force_utf8_rec` caller to pass `_as_bytes=True` and change this `elif` case into:
```
    elif _as_bytes and is_unicode(input_obj):
	return _force_primitive(input_obj, highlight)
    elif not _as_bytes and not is_unicode(input_obj):
        return _force_primitive(input_obj, highlight)
```

Background info: as I think you are already well aware of `force_utf8` **must** effectively return `bytes` (`b"bla"`) and `force_unicode` must correspondingly return `unicode` (`u"bla"` ) no matter what the platform default string type is. The same applies for the contents returned by the `force_X_rec` versions.